### PR TITLE
Implement setter for document.domain

### LINKED ISSUE #934: Implement document.domain
This is related to #633.
document.domain changes the origin of a document. Any current and coming optimizations based on the assumption that the origin of a document not changing over time like the recent [Make sandboxed iframes run with different script tasks](https://github.com/mozilla/servo/commit/96403247218ce628a39b954c361d0707d0d0f97d) may require removing these optimizations later.

For that reason, I feel it'd be important to implement document.domain early and start using it in Servo tests to make sure Servo is actually running the web and not an idealized version

For the specific case of sandbox iframes and document.domain, it's been discussed [in standards](https://www.w3.org/Bugs/Public/show_bug.cgi?id=23040), there was agreement among the few who participated in the discussion and a solution has been [implemented in Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=907892) (poison document.domain setter in sandboxed iframes regardless of `allow-same-origin`).
Assuming this sticks, this particular optimization will remain.

### PR DESCRIPTION
**Pull request by asajeffrey**
This PR implements the setter for `document.domain`.

It builds on #15438 and #15478, only the last commit is part of this PR.

It includes tests for similar-origin security checks.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #934.
- [X] There are tests for these changes.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15536)

### CODE REVIEW

**Comment by asajeffrey** [comment]
cc @jdm @Ms2ger

**Comment by asajeffrey** [comment]
Updated the wpt manifest.

**Comment by asajeffrey** [comment]
IRC conversation with @jdm: http://logs.glob.uno/?c=mozilla%23servo&s=24+Feb+2017&e=24+Feb+2017#c619500

TL;DR: closed #15478 in favour of merging with this PR, so we can include wpt tests.

**Comment by asajeffrey** [comment]
The spec for document.domain changed: https://github.com/whatwg/html/commit/03bbc6e92b6b8c9735b4565722916d49fe3eab61

I added some more wpt tests.

@emilio, @Ms2ger and @jdm: This is ready for review, since #15478 was closed in favour of this PR.

**Comment by asajeffrey** [comment]
The spec for `document.open()` changed: https://github.com/whatwg/html/pull/2288

IRC chat with @Ms2ger: http://logs.glob.uno/?c=mozilla%23servo&s=27+Feb+2017&e=27+Feb+2017#c620440 - now that there's a wpt test for `document.open()` we should have sufficient test coverage.

File: components/script/dom/webidls/Document.webidl [hunk 1]
```diff
@@ -81,7 +81,7 @@ partial /*sealed*/ interface Document {
   // resource metadata management
   [/*PutForwards=href, */Unforgeable]
   readonly attribute Location? location;
-  readonly attribute DOMString domain;
+  [Throws] attribute DOMString domain;
```
Comment (asajeffrey) [inline components/script/dom/webidls/Document.webidl]: Oh, I didn't know there was such a thing! Yes, it should.

Comment (asajeffrey) [inline components/script/dom/webidls/Document.webidl]: Done.

**Comment by asajeffrey** [comment]
Fixes #15233.

**Comment by asajeffrey** [comment]
Squashed down into two commits. The first used to be #15478, the second is the implementation of the setter for document.domain.

File: tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/common.js [hunk 2]
```diff
@@ -195,8 +195,8 @@ function queryAnchor(url, callback, referrer_policy) {
     var url_with_params = url + "&id=" + id + "&tagAttrs=" + JSON.stringify(referrer_policy);
     var iframe = appendIframeToBody(url_with_params);
     iframe.addEventListener("load", function listener() {
-      if ((iframe.contentWindow !== null) &&
-          (iframe.contentWindow.location.toString() === url_with_params)) {
+      if ((iframe.contentDocument !== null) &&
+          (iframe.contentDocument.location.toString() === url_with_params)) {
```
Comment (nox) [inline tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/common.js]: Why is `iframe.contentWindow.location` not ok but `iframe.contentDocument.location` is?

File: components/script/dom/document.rs [hunk 3]
```diff
@@ -1988,6 +1989,53 @@ impl LayoutDocumentHelpers for LayoutJS<Document> {
...
+        }
+
+        // Step 4.3
+        if is_pub_domain(host) {
+            return false;
+        }
+    }
+
+    // Step 5
+    return true;
```
Comment (nox) [inline components/script/dom/document.rs]: Nit: remove `return`.

File: components/script/dom/document.rs [hunk 4]
```diff
@@ -2489,6 +2537,38 @@ impl DocumentMethods for Document {
         }
     }
 
+    // https://html.spec.whatwg.org/multipage/#relaxing-the-same-origin-restriction
```
Comment (nox) [inline components/script/dom/document.rs]: Nit: https://html.spec.whatwg.org/multipage/#dom-document-domain

File: components/script/dom/document.rs [hunk 5]
```diff
@@ -2489,6 +2537,38 @@ impl DocumentMethods for Document {
...
+        // Step 1.
+        if !self.has_browsing_context {
+            return Err(Error::Security);
+        }
+
+        // TODO: Step 2. "If this Document object's active sandboxing
+        // flag set has its sandboxed document.domain browsing context
+        // flag set, then throw a "SecurityError" DOMException."
+
+        // Step 3-4.
```
Comment (nox) [inline components/script/dom/document.rs]: Nit: plural. :)

File: components/script/dom/document.rs [hunk 6]
```diff
@@ -2489,6 +2537,38 @@ impl DocumentMethods for Document {
...
+            None => return Err(Error::Security),
+        };
+
+        // Step 5
+        if !is_a_registrable_domain_suffix_of_or_is_equal_to(&*value, effective_domain) {
+            return Err(Error::Security);
+        }
+
+        // Step 6
+        // Slightly annoying to have to parse twice.
```
Comment (nox) [inline components/script/dom/document.rs]: Make `is_a_registrable_domain_suffix_of_or_is_equal_to` return `Fallible<Host>`.

File: components/script/dom/document.rs [hunk 7]
```diff
@@ -3396,10 +3476,10 @@ impl DocumentMethods for Document {
 
         let entry_responsible_document = GlobalScope::entry().as_window().Document();
 
-        // This check should probably be same-origin-domain
+        // This check is same-origin not same-origin-domain.
```
Comment (nox) [inline components/script/dom/document.rs]: I don't understand that comment.

**Review by nox — CHANGES_REQUESTED** [review]
Many comments, many nits, but good work.

File: components/net_traits/pub_domains.rs [hunk 8]
```diff
@@ -146,6 +146,11 @@ pub fn is_reg_domain(domain: &str) -> bool {
 /// Returns None if the URL has no host name.
 /// Returns the registered suffix for the host name if it is a domain.
 /// Leaves the host name alone if it is an IP address.
-pub fn reg_host<'a>(url: &'a ServoUrl) -> Option<&'a str> {
-    url.domain().map(reg_suffix).or(url.host_str())
+pub fn reg_host<'a>(url: &'a ServoUrl) -> Option<Host> {
```
Comment (asajeffrey) [inline components/net_traits/pub_domains.rs]: Done.

File: components/script/dom/history.rs [hunk 9]
```diff
@@ -65,7 +64,7 @@ impl HistoryMethods for History {
         } else if delta < 0 {
             TraversalDirection::Back(-delta as usize)
         } else {
-            self.window.Location().Reload();
+            self.window.Location().reload_without_origin_check();
```
Comment (asajeffrey) [inline components/script/dom/history.rs]: On #whatwg they pointed out that history can't be accessed cross-origin, so once we have XOWs this point is moot. I'll put it back to `Reload`.

File: components/script/dom/location.rs [hunk 10]
```diff
@@ -43,11 +45,27 @@ impl Location {
...
+    // https://html.spec.whatwg.org/multipage/#dom-location-reload
+    pub fn reload_without_origin_check(&self) {
+        self.window.load_url(self.get_url(), true, true, None);
+    }
 }
 
 impl LocationMethods for Location {
     // https://html.spec.whatwg.org/multipage/#dom-location-assign
     fn Assign(&self, url: USVString) -> ErrorResult {
+        // Note: no call to self.check_same_origin_domain()
```
Comment (asajeffrey) [inline components/script/dom/location.rs]: Good question. I must have dreamed the no-security-check here.

File: components/script/dom/window.rs [hunk 11]
```diff
@@ -42,7 +42,7 @@ use dom::location::Location;
 use dom::mediaquerylist::{MediaQueryList, WeakMediaQueryListVec};
 use dom::messageevent::MessageEvent;
 use dom::navigator::Navigator;
-use dom::node::{Node, from_untrusted_node_address, window_from_node, NodeDamage};
+use dom::node::{Node, from_untrusted_node_address, document_from_node, window_from_node, NodeDamage};
```
Comment (asajeffrey) [inline components/script/dom/window.rs]: Done.

File: components/script/dom/window.rs [hunk 12]
```diff
@@ -528,7 +528,20 @@ impl WindowMethods for Window {
...
+                // Step 6.
+                let container_doc = document_from_node(container);
+                let current_doc = GlobalScope::current().as_window().Document();
+                if current_doc.origin().same_origin_domain(container_doc.origin()) {
+                    // Step 7.
+                    return Some(Root::from_ref(container));
+                }
+            }
+        }
+        None
```
Comment (asajeffrey) [inline components/script/dom/window.rs]: Done.

File: tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/common.js (hunk 2 above)
Comment (asajeffrey) [inline tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/common.js]: Because `iframe.contentWindow` will succeed, even if the current document is cross-origin, but `iframe.contentWindow.location.toString()` will then throw a security error.

**Review by asajeffrey — COMMENTED** [review]
Thanks for the careful review!

**Comment by asajeffrey** [comment]
Pushed again to get appveyor to rebuild.

File: components/net_traits/pub_domains.rs [hunk 13]
```diff
@@ -146,11 +146,13 @@ pub fn is_reg_domain(domain: &str) -> bool {
...
+pub fn reg_host(url: &ServoUrl) -> Option<Host> {
     match url.origin() {
-        ImmutableOrigin::Tuple(_, Host::Domain(domain), _) => Some(Host::Domain(String::from(reg_suffix(&*domain)))),
-        ImmutableOrigin::Tuple(_, Host::Ipv4(address), _) => Some(Host::Ipv4(address)),
-        ImmutableOrigin::Tuple(_, Host::Ipv6(address), _) => Some(Host::Ipv6(address)),
+        ImmutableOrigin::Tuple(_, host, _) => if let Host::Domain(domain) = host {
+            Some(Host::Domain(String::from(reg_suffix(&*domain))))
+        } else {
+            Some(host)
+        },
```
Comment (nox) [inline components/net_traits/pub_domains.rs]: This is worse than before or what I suggested and the formatting is wrong. Don't nest an `if let` expression inside a match arm without enclosing it in braces.

File: components/script/dom/document.rs [hunk 14]
```diff
@@ -1990,50 +1990,52 @@ impl LayoutDocumentHelpers for LayoutJS<Document> {
 }
 
 // https://html.spec.whatwg.org/multipage/#is-a-registrable-domain-suffix-of-or-is-equal-to
-fn is_a_registrable_domain_suffix_of_or_is_equal_to(host_suffix_string: &str, original_host: Host) -> bool {
+// The spec says to return a bool, we actually return an Option<Host> containing
+// the parsed host in the successful case, to avoid having to re-parse the host.
+fn is_a_registrable_domain_suffix_of_or_is_equal_to(host_suffix_string: &str, original_host: Host) -> Option<Host> {
```
Comment (nox) [inline components/script/dom/document.rs]: Rename the method to something that doesn't sound like a boolean.

File: components/script/dom/document.rs [hunk 15]
```diff
@@ -2548,23 +2550,20 @@ impl DocumentMethods for Document {
...
         };
 
         // Step 5
-        if !is_a_registrable_domain_suffix_of_or_is_equal_to(&*value, effective_domain) {
-            return Err(Error::Security);
-        }
+        let host = match is_a_registrable_domain_suffix_of_or_is_equal_to(&*value, effective_domain) {
+            None => return Err(Error::Security),
+            Some(host) => host,
+        };
```
Comment (nox) [inline components/script/dom/document.rs]:
```rust
let host = is_a_registrable_domain_suffix_of_or_is_equal_to(&*value, effective_domain).ok_or(Error::Security)?;
```

**Review by nox — CHANGES_REQUESTED** [review]

File: components/net_traits/pub_domains.rs [hunk 16]
```diff
@@ -146,6 +146,12 @@ pub fn is_reg_domain(domain: &str) -> bool {
...
-pub fn reg_host<'a>(url: &'a ServoUrl) -> Option<&'a str> {
-    url.domain().map(reg_suffix).or(url.host_str())
+pub fn reg_host(url: &ServoUrl) -> Option<Host> {
+    match url.origin() {
+        ImmutableOrigin::Tuple(_, host, _) => match host {
+            Host::Domain(domain) => Some(Host::Domain(String::from(reg_suffix(&*domain)))),
+            _ => Some(host),
+        },
+        ImmutableOrigin::Opaque(_) => None,
+    }
```
Comment (nox) [inline components/net_traits/pub_domains.rs]: Still the same issue?

Either just copy/paste my original comment, or wrap the nested match in braces.

```rust
match url.origin() {
    ImmutableOrigin::Tuple(_, Host::Domain(domain), _) => Some(Host::Domain(String::from(reg_suffix(&*domain)))),
    ImmutableOrigin::Tuple(_, ip, _) => Some(ip),
    ImmutableOrigin::Opaque(_) => None,
}
```

Comment (nox) [inline components/net_traits/pub_domains.rs]: Or a mix of the two.

```rust
Nit: you can merge the two arms above together AFAICT.

```rust
match url.origin() {
    ImmutableOrigin::Tuple(_, Host::Domain(domain), _) => {
        Some(Host::Domain(String::from(reg_suffix(&*domain))))
    },
    ImmutableOrigin::Tuple(_, ip, _) => Some(ip),
    ImmutableOrigin::Opaque(_) => None,
}
```

**Comment by nox** [comment]
r=me when the tests pass.

**Review by nox — APPROVED** [review]

**Comment by nox** [comment]
The intermittent crash seems to be a race condition when writing the `alert()` message, please add a security check to it so that it will throw `SecurityError` as the spec mandates in the test, thus avoiding the intermittent crash entirely.

**Comment by asajeffrey** [comment]
Added #15962.

**Comment by asajeffrey** [comment]
Squashed down to two commits.

@bors-servo: r=nox
